### PR TITLE
Normalize SQL IN(?, ?, ...) statements to "in(?)" to reduce cardinality of db.statement attribute

### LINKED ISSUE #10442: Normalize "IN(?, ?, ...)" substrings in db.statement span attribute to "IN(?)"
### Is your feature request related to a problem? Please describe.

Hello, we're looking to use `db.statement` as an attribute in span metrics to get RED metrics for each query

Currently, SQL queries using `IN` with a variable number of values will each generate a different `db.statement`, varying only by the number of `?` in the `IN` part. That can cause high cardinality span metrics and makes the metrics difficult to analyze

### Describe the solution you'd like

Replace all `IN(...)` statements in SQL `db.statement`s with a constant value like `IN( ? )`

### Describe alternatives you've considered

Using the otel collector to normalize the query string

### Additional context

Let me know if this would need to first be approved as a semantic convention

### PR DESCRIPTION
**Pull request by swar8080**
Went for a simple implementation, but let me know if any of the following would be preferred:
- A way to turn this new behaviour on/off
- Decoupling sanitizing and normalizing sql statements into separate classes/abstractions
- Implementing normalization as part of the jflex parsing to avoid another pass through the query string. I wouldn't have the bandwidth to implement that though

Closes  #10442

### CODE REVIEW

**Review by zeitlinger — APPROVED** [review]

**Review by trask — COMMENTED** [review]
can this be done without using a regular expression? (my concern is performance)

**Comment by zeitlinger** [comment]
> can this be done without using a regular expression? (my concern is performance)

it's precompiled - and it doesn't use pathological patterns

**Comment by swar8080** [comment]
my bad, this is actually causing a stack overflow for an IN statement with 2000 values

ill see if there's a different approach

File: instrumentation-api-incubator/src/main/jflex/SqlSanitizer.jflex [hunk 1]
```diff
@@ -52,6 +54,9 @@ WHITESPACE          = [ \t\r\n]+
   // max length of the sanitized statement - SQLs longer than this will be trimmed
   static final int LIMIT = 32 * 1024;
 
+  private static final Pattern IN_STATEMENT_PATTERN = Pattern.compile("\\sin\\s*\\(\\s*\\?[\\s?,]*?\\)", Pattern.CASE_INSENSITIVE);
```
Comment (laurit) [inline instrumentation-api-incubator/src/main/jflex/SqlSanitizer.jflex:57]: I think this also matches inputs like ` in (?,,,???)` perhaps using `"(\\sin\\s*)\\(\\s*\\?\\s*(,\\s*\\?\\s*)*\\)"` and replacing with `"$1(?)"` would be better. These regular expressions are hard to parse, maybe we should try to document them to make them easier to understand?

Comment (laurit) [reply instrumentation-api-incubator/src/main/jflex/SqlSanitizer.jflex:58]: > Unfortunately the (,\\s*\\?\\s*)* part of that pattern causes a stack overflow for IN statements with many values

using a possessive quantifier should fix this, try `"(\\sin\\s*)\\(\\s*\\?\\s*(,\\s*\\?\\s*)*+\\)"`

Comment (swar8080) [reply instrumentation-api-incubator/src/main/jflex/SqlSanitizer.jflex]: thanks, that solves the problem

**Comment by laurit** [comment]
closing and reopening to trigger checks

File: instrumentation-api-incubator/src/main/jflex/SqlSanitizer.jflex [hunk 2]
```diff
@@ -52,6 +54,10 @@ WHITESPACE          = [ \t\r\n]+
   // max length of the sanitized statement - SQLs longer than this will be trimmed
   static final int LIMIT = 32 * 1024;
 
+  // Match on "IN(?, ?, ...)"
+  private static final Pattern IN_STATEMENT_PATTERN = Pattern.compile("(\\sin\\s*)\\(\\s*\\?\\s*(,\\s*\\?\\s*)*+\\)", Pattern.CASE_INSENSITIVE);
+  private static final String IN_STATEMENT_NORMALIZED = " in(?)";
```
Comment (laurit) [inline instrumentation-api-incubator/src/main/jflex/SqlSanitizer.jflex:59]:
```suggestion
  private static final String IN_STATEMENT_NORMALIZED = "$1(?)";
```
Sanitizer does not change case or remove whitespace from the original query. Lets keep `in` as it was in the original query. `longInStatementDoesntCauseStackOverflow` will break after this change as there is a space between in and ( that is currently removed.

Comment (swar8080) [reply instrumentation-api-incubator/src/main/jflex/SqlSanitizer.jflex]: Got it, I updated it to preserve case and whitespace, and updated the test cases to check for that. I didn't add a test case for more than one space between `IN` and `(` since strings like `IN   (?)` get sanitized to `IN (?)`

Also switched to a non-capturing group for matching on the part in-between the brackets as a small optimization

**Review by laurit — APPROVED** [review]

**Comment by swar8080** [comment]
HI @laurit, let me know if anything else should be done here

Would be great to have this in 2.2.0
